include/mbedtls/cipher.h

```diff
@@ -218,125 +231,133 @@ typedef struct {
  * Generic cipher context.
  */
 typedef struct {
-    /** Information about the associated cipher */
+    /** Information about the associated cipher. */
     const mbedtls_cipher_info_t *cipher_info;
 
-    /** Key length to use */
+    /** Key length to use. */
     int key_bitlen;
 
-    /** Operation that the context's key has been initialised for */
+    /** Operation that the key of the context has been 
+	 * initialized for. 
+	 */
     mbedtls_operation_t operation;
 
 #if defined(MBEDTLS_CIPHER_MODE_WITH_PADDING)
-    /** Padding functions to use, if relevant for cipher mode */
+    /** Padding functions to use, if relevant for 
+	 * the specific cipher mode. 
+	 */
     void (*add_padding)( unsigned char *output, size_t olen, size_t data_len );
     int (*get_padding)( unsigned char *input, size_t ilen, size_t *data_len );
 #endif
 
-    /** Buffer for data that hasn't been encrypted yet */
+    /** Buffer for data that has not been encrypted. */
```

Comment:
I disagree. The data in the buffer has either been encrypted, or has not been encrypted. The fact that we plan to encrypt it does not make a difference here, unless I am missing something.

Reply:
Some algorithms can only process data one block at a time. Let's take the example of AES-CBC encryption, which has 16-byte blocks. The caller can request the encryption of, say, 17 bytes with `mbedtls_cipher_update`. The call to `mbedtls_cipher_update` encrypts the first 16 bytes, and leaves the remaining byte in `unprocessed_data`. This extra byte will stay in `unprocessed_data` until a later call to `mbedtls_cipher_update`, when at least 15 more bytes of input have accumulated, at which point the accumulated input gets encrypted. (And there are special rules for `mbedtls_cipher_final` to encrypt less than 16 bytes at the end.)

The whole point of `unprocessed_data` is that it contains data that has been submitted for encryption, and will be encrypted by a subsequent call to a function using this context. The fact that we plan to encrypt it is the whole point of this buffer. Describing it as “data that has not been encrypted” rather than “data that has not been encrypted yet” makes it look like this is data that remains unencrypted, which is a different concept.

Now that I think of it, this buffer can also contain input that has not been decrypted yet, in a decryption operation. So the former description was wrong too, it should have been “data that has not been processed yet” or more precisely “input that has not been processed yet”.

Reply:
Thank you for the detailed explanation Gilles! It is now clear why this distinction is important.